### PR TITLE
[zero] memtracer to record cuda memory usage of model data and overall system

### PR DESCRIPTION
**Pull request by feifeibear**
Design a tracer to collect memory statistics during training.
statistics include 
1. model data memory usage (we use an explicit interface to trace allocation, release and moving)
2. system allocated cuda memory.

### CODE REVIEW

File: colossalai/utils/memory_tracer/allocator.py [hunk 1]
```diff
@@ -1,60 +1,19 @@
...
-from colossalai.utils.commons.singleton_meta import SingletonMeta
-from colossalai.zero.sharded_param import ShardedTensor
+from colossalai.utils.memory_tracer.model_data_memtracer import ModelDataTracer
 
-from typing import Union
 
+def col_move_to_cpu(t: torch.Tensor):
+    assert isinstance(t, torch.Tensor)
+    if t.device.type == 'cpu':
+        return
```
Comment (ver217) [inline colossalai/utils/memory_tracer/allocator.py:8]: Should we rename this function?

File: colossalai/utils/memory_tracer/memstats_collector.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        """
+        returns cuda usage of model data and overall cuda usage.
+        """
+        sampling_cnt = self._sampling_cnter.sampling_cnt
+        if len(self._model_data_cuda) < sampling_cnt:
+            raise RuntimeError
+        return (self._model_data_cuda[sampling_cnt], self._overall_cuda[sampling_cnt])
+
+    def reset(self) -> None:
+        self._sampling_cnter.reset()
```
Comment (ver217) [inline colossalai/utils/memory_tracer/memstats_collector.py:71]: Should we provide a method to reset `_model_data_cuda` and `_overall_cuda`? Their sizes are always increasing now.

Comment (feifeibear) [reply colossalai/utils/memory_tracer/memstats_collector.py:71]: I will rename the function as reset_sampling_cnt(self) and add a clear function to reset the list.

File: colossalai/zero/sharded_model/sharded_model_v2.py [hunk 3]
```diff
@@ -59,8 +61,16 @@ def __init__(self,
                 if self.shard_param:
                     self.shard_strategy.shard([param.col_attr.data])
 
+        # Init Memory Statistics Collector
+        self._use_memory_tracer = use_memory_tracer
+        if self._use_memory_tracer:
+            self._memstats_collector = MemStatsCollector()
```
Comment (ver217) [inline colossalai/zero/sharded_model/sharded_model_v2.py:67]: This collector is designed only for `ShardedModel` or for all models? Should we manage it in `engine`?

Comment (feifeibear) [reply colossalai/zero/sharded_model/sharded_model_v2.py:67]: Now just used for ShardedModel. It highly depends on the zerohook.

File: tests/test_zero_data_parallel/test_shard_model_v2.py [hunk 4]
```diff
@@ -20,12 +20,14 @@
...
 from colossalai.zero.sharded_model.utils import col_model_deepcopy
+from colossalai.utils import get_current_device
 
 
 def run_dist(rank, world_size, port, use_zero_init_ctx, enable_autocast):
     colossalai.launch(config=CONFIG, rank=rank, world_size=world_size, host='localhost', port=port, backend='nccl')
 
-    test_models = ['repeated_computed_layers', 'resnet18', 'bert']
+    # test_models = ['repeated_computed_layers', 'resnet18', 'bert']
+    test_models = ['bert']
```
Comment (ver217) [inline tests/test_zero_data_parallel/test_shard_model_v2.py:30]: Why just test BERT here?

Comment (feifeibear) [reply tests/test_zero_data_parallel/test_shard_model_v2.py:30]: revert it back.

**Review by ver217 — APPROVED** [review]

**Pull request merged**